tdrs-backend/tdpservice/parsers/parse.py

```diff
@@ -61,27 +41,71 @@ def parse_datafile(datafile):
         errors['document'] = ['Section does not match.']
         return errors
 
-    line_errors = parse_datafile_lines(datafile, program_type, section)
+    line_errors = parse_datafile_lines(rawfile, program_type, section)
 
     errors = errors | line_errors
 
     return errors
 
 
-def parse_datafile_lines(datafile, program_type, section):
+def store_record(unsaved_records, record, model):
+    """Store record in dictionary for later processing."""
+    if record:
+        if model not in unsaved_records:
+            unsaved_records.update({model: [record]})
```

Comment:
`store_record`, as well as `evaluate_trailer` below, break both the purity and immutability principles of functional programming because they modify one of the passed-in parameters (`unsaved_records` in this case). A slightly better solution might be to build a new `unsaved_records` object based on a copy of the original and return it, overriding the value in the calling function.

```python
def store_record(unsaved_records, record, model):
    new_records = unsaved_records.copy()
    if record:
        new_records.get(model, []).append(record)  # might work as a 1-line replacement of lines 55-58
    return new_records


## then, inside `parse_datafile_lines`

unsaved_records = store_record(unsaved_records, record, model)  # override the previously set value
```